keystore/v2/keystore/api/key.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/cossacklabs/acra/keystore/v2/keystore/asn1"
+)
+
+// Errors returned by Key methods:
+var (
+	ErrFormatDuplicated    = errors.New("KeyStore: key format used multiple times")
+	ErrFormatMissing       = errors.New("KeyStore: key format not available")
+	ErrKeyNotExist         = errors.New("KeyStore: no key with such seqnum")
+	ErrNoKeyData           = errors.New("KeyStore: no key data")
+	ErrInvalidFormat       = errors.New("KeyStore: invalid key format")
+	ErrInvalidState        = errors.New("KeyStore: invalid state transition")
+	ErrInvalidCryptoperiod = errors.New("KeyStore: invalid key cryptoperiod")
+)
+
+// Key in a KeyRing.
+type Key interface {
+	// Seqnum returns sequential number of this key in the key ring.
+	Seqnum() (int, error)
+	// State of the key right now.
+	State() (KeyState, error)
+
+	// ValidSince returns the time before which the key cannot be used.
+	ValidSince() (time.Time, error)
+	// ValidUntil returns the time since which the key should not be used.
+	ValidUntil() (time.Time, error)
+
+	// Formats available for this key.
+	Formats() ([]KeyFormat, error)
+	// PublicKey data in given format, if available.
+	PublicKey(format KeyFormat) ([]byte, error)
+	// PrivateKey data in given format, if available.
+	PrivateKey(format KeyFormat) ([]byte, error)
+	// SymmetricKey data in given format, if available.
+	SymmetricKey(format KeyFormat) ([]byte, error)
+}
+
+// MutableKey in a KeyRing.
+type MutableKey interface {
+	Key
+
+	// SetState changes key State to the given one, if allowed.
+	SetState(newState KeyState) error
+
+	// SetCurrent makes this key current in its key ring.
+	// Does nothing if the key is already current.
+	SetCurrent() error
```

Comment:
> as I understand, it requires from MutableKey implementation to have pointer to KeyRing implementation to be able to setup itself as current key, yes?

Yeah, you’re right. A key keeps a pointer to its parent key ring as well as its own sequential number so that it’s possible to tell which key it is. This is used not only to set the current key, but to request key data as well because it’s actually stored in the key ring, not in the key.

I don’t really like this indirection. (And it used to be even more silly, like the key keeping an actual _pointer_ into data stored inside the key ring — which had to be refreshed in all existing keys when a new key is added to the ring.) However, my attempts at removing it ended up in just abolishing the whole `Key` interface altogether and interacting with keys via the key ring. Maybe that would be a better, like

```go
type KeyRing interface {
        CurrentKey() (int, error)
        SetCurrentKey(seqnum int) error

        AddKey(key KeyDescription) (int, error)

        PublicKey(seqnum int, format KeyFormat) ([]byte, error)
        PrivateKey(seqnum int, format KeyFormat) ([]byte, error)
}
```

I’m not sure it’s good, but I’m not sure we’d need a separate `Key` interface either.

Well, now that you have seen the code, let’s discuss it offline.

Reply:
@Lagovas, I’ve coded a simplified API like shown above which does not rely on a separate `Key` instances to access keys in. All key accesses are performed via a `KeyRing` object. Here’s a [diff against this PR](https://github.com/ilammy/acra/compare/keystore-v2...simple-key-api) (changes only), and a [complete diff](https://github.com/cossacklabs/acra/compare/keystore-v2...ilammy:simple-key-api) (easier to read).

Could you please give it a look? I’ll include it into this PR if that’s fine by you.

(I’ve simply renamed the interfaces and fixed up the code so that it’s clear what changes are made. All those awkward `KeyAccess` interfaces are unlikely to be useful on their own so they can be merged into `KeyRing` directly.)

Reply:
I’ve pushed the commit which moves all key access API into `KeyRing` interface.